In [14]:
import pandas as pd #first we'll need module pandas to work with the dataframes
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
import statsmodels.formula.api as smf
from datetime import datetime
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn import metrics


In [93]:
weather_data = pd.read_csv('DAPR.csv')
weather_data.count()


STATION            11872
NAME               11872
LATITUDE           11872
LONGITUDE          11872
ELEVATION          11872
DATE               11872
DAPR                 235
DAPR_ATTRIBUTES      235
MDPR                 233
MDPR_ATTRIBUTES      233
dtype: int64

STATION            11872
NAME               11872
LATITUDE           11872
LONGITUDE          11872
ELEVATION          11872
DATE               11872
DAPR                 235
DAPR_ATTRIBUTES      235
MDPR                 233
MDPR_ATTRIBUTES      233
dtype: int64

In [36]:
weather_data.dropna(subset=['DAPR'])

STATION                             NAME   LATITUDE  LONGITUDE  \
27     US1NJBG0043            MAYWOOD 0.2 SW, NJ US  40.900010 -74.066410   
42     US1NJBG0043            MAYWOOD 0.2 SW, NJ US  40.900010 -74.066410   
126    US1NJBG0043            MAYWOOD 0.2 SW, NJ US  40.900010 -74.066410   
161    US1NJES0031           MONTCLAIR 0.7 N, NJ US  40.835343 -74.208967   
168    US1NJES0031           MONTCLAIR 0.7 N, NJ US  40.835343 -74.208967   
...            ...                              ...        ...        ...   
11739  US1NJES0019  WEST CALDWELL TWP 1.3 NE, NJ US  40.861500 -74.277500   
11773  US1NJES0019  WEST CALDWELL TWP 1.3 NE, NJ US  40.861500 -74.277500   
11857  US1NJES0019  WEST CALDWELL TWP 1.3 NE, NJ US  40.861500 -74.277500   
11869  US1NJES0019  WEST CALDWELL TWP 1.3 NE, NJ US  40.861500 -74.277500   
11871  US1NJES0019  WEST CALDWELL TWP 1.3 NE, NJ US  40.861500 -74.277500   

       ELEVATION        DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
27          24.7  2021-08-06   3.0             ,,N  0.00             ,,N  
42          24.7  2021-08-28   8.0             ,,N  0.99             ,,N  
126         24.7  2021-11-27   2.0             ,,N  0.18             ,,N  
161         96.0  2021-01-01   3.0             ,,N  0.32             ,,N  
168         96.0  2021-01-11   4.0             ,,N  0.00             ,,N  
...          ...         ...   ...             ...   ...             ...  
11739       81.4  2021-07-20   4.0             ,,N  1.60             ,,N  
11773       81.4  2021-08-30   6.0             ,,N  1.37             ,,N  
11857       81.4  2021-11-30   6.0             ,,N  0.19             ,,N  
11869       81.4  2021-12-17   6.0             ,,N  0.07             ,,N  
11871       81.4  2021-12-31  13.0             ,,N  0.85             ,,N  

[235 rows x 10 columns]

In [94]:
weather_data.shape

(11872, 10)

Data size is 11872. 

In [95]:
weather_data['DATE'] = weather_data['DATE'].astype('datetime64')

In [96]:
weather_data.DATE

0       2021-06-24
1       2021-06-25
2       2021-06-26
3       2021-06-27
4       2021-06-28
           ...    
11867   2021-12-10
11868   2021-12-11
11869   2021-12-17
11870   2021-12-18
11871   2021-12-31
Name: DATE, Length: 11872, dtype: datetime64[ns]

In [97]:
weather_data.dtypes

STATION                    object
NAME                       object
LATITUDE                  float64
LONGITUDE                 float64
ELEVATION                 float64
DATE               datetime64[ns]
DAPR                      float64
DAPR_ATTRIBUTES            object
MDPR                      float64
MDPR_ATTRIBUTES            object
dtype: object

fliter the ONlY NYC Data. 

In [101]:
index = (weather_data['LATITUDE'] > 40.495992) & (weather_data['LATITUDE'] < 40.915568) &\
    (weather_data['LONGITUDE'] > -74.257159) & (weather_data['LONGITUDE'] < -73.699215) 
weather_data = weather_data.loc[index]

In [102]:
weather_data.count()

STATION            6992
NAME               6992
LATITUDE           6992
LONGITUDE          6992
ELEVATION          6992
DATE               6992
DAPR                147
DAPR_ATTRIBUTES     147
MDPR                146
MDPR_ATTRIBUTES     146
dtype: int64

Final Size is the 6992.

We can see that there is only 147/6992 records of any rain observed in New York City. There is not sufficient data for us to analyze.

In [43]:
df=weather_data

In [46]:
df

STATION                      NAME  LATITUDE  LONGITUDE  ELEVATION  \
0      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
1      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
2      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
3      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
4      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
...            ...                       ...       ...        ...        ...   
11208  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11209  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11210  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11211  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11212  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   

            DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
0     2021-06-24   NaN             NaN   NaN             NaN  
1     2021-06-25   NaN             NaN   NaN             NaN  
2     2021-06-26   NaN             NaN   NaN             NaN  
3     2021-06-27   NaN             NaN   NaN             NaN  
4     2021-06-28   NaN             NaN   NaN             NaN  
...          ...   ...             ...   ...             ...  
11208 2021-12-27   NaN             NaN   NaN             NaN  
11209 2021-12-28   NaN             NaN   NaN             NaN  
11210 2021-12-29   NaN             NaN   NaN             NaN  
11211 2021-12-30   NaN             NaN   NaN             NaN  
11212 2021-12-31   NaN             NaN   NaN             NaN  

[6992 rows x 10 columns]

In [84]:
df[df["DAPR"]==0]

Empty DataFrame
Columns: [STATION, NAME, LATITUDE, LONGITUDE, ELEVATION, DATE, DAPR, DAPR_ATTRIBUTES, MDPR, MDPR_ATTRIBUTES]
Index: []

In [45]:
df.describe()

LATITUDE    LONGITUDE    ELEVATION        DAPR        MDPR
count  6992.000000  6992.000000  6992.000000  147.000000  146.000000
mean     40.766997   -74.054247    26.496424    6.639456    1.153082
std       0.099150     0.133221    28.933622    8.133261    1.655622
min      40.542096   -74.238792     0.800000    2.000000    0.000000
25%      40.684600   -74.152000     5.500000    3.000000    0.075000
50%      40.778980   -74.066410    19.800000    4.000000    0.465000
75%      40.848094   -73.969250    29.000000    6.000000    1.480000
max      40.914670   -73.810340   108.500000   53.000000    8.250000

In [87]:
df['NAME'].unique()

array(['MAYWOOD 0.2 SW, NJ US', 'MONTCLAIR 0.7 N, NJ US',
       'NEW YORK 8.8 N, NY US', 'MONTCLAIR 0.7 NNE, NJ US',
       'TENAFLY 1.3 W, NJ US', 'LITTLE FALLS TWP 0.2 NE, NJ US',
       'BROOKLYN 3.1 NW, NY US', 'HOWARD BEACH 0.4 NNW, NY US',
       'STATEN ISLAND 2.6 N, NY US', 'LAGUARDIA AIRPORT, NY US',
       'QUEENS 3.7 N, NY US',
       'NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US',
       'MATTHEWS PALMER PLAYGROUND NEW YORK 6.8 NNW, NY US',
       'CEDAR GROVE TWP 0.4 W, NJ US', 'NORTH ARLINGTON 0.7 WNW, NJ US',
       'KEARNY 1.7 NW, NJ US', 'PALISADES PARK 0.2 WNW, NJ US',
       'HARRISON, NJ US', 'CARTERET 0.6 WSW, NJ US',
       'LITTLE FALLS TWP 0.5 WNW, NJ US', 'STATEN ISLAND 1.4 SE, NY US',
       'STATEN ISLAND 4.5 SSE, NY US', 'NY CITY CENTRAL PARK, NY US',
       'JACKSON HEIGHTS 0.3 WSW, NY US', 'QUEENS 3.4 NNW, NY US',
       'TETERBORO AIRPORT, NJ US'], dtype=object)

In [66]:
df[df["NAME"]=='MAYWOOD 0.2 SW, NJ US']

STATION                   NAME  LATITUDE  LONGITUDE  ELEVATION  \
0    US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
1    US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
2    US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
3    US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
4    US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
..           ...                    ...       ...        ...        ...   
156  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
157  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
158  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
159  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
160  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   

          DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
0   2021-06-24   NaN             NaN   NaN             NaN  
1   2021-06-25   NaN             NaN   NaN             NaN  
2   2021-06-26   NaN             NaN   NaN             NaN  
3   2021-06-27   NaN             NaN   NaN             NaN  
4   2021-06-28   NaN             NaN   NaN             NaN  
..         ...   ...             ...   ...             ...  
156 2021-12-27   NaN             NaN   NaN             NaN  
157 2021-12-28   NaN             NaN   NaN             NaN  
158 2021-12-29   NaN             NaN   NaN             NaN  
159 2021-12-30   NaN             NaN   NaN             NaN  
160 2021-12-31   NaN             NaN   NaN             NaN  

[161 rows x 10 columns]

In [72]:
df[df["NAME"]=='NEW YORK 8.8 N, NY US']

STATION                   NAME  LATITUDE  LONGITUDE  ELEVATION  \
453  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
454  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
455  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
456  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
465  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
466  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
483  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
487  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
491  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
493  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
494  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
495  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
496  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
507  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   
508  US1NYNY0074  NEW YORK 8.8 N, NY US  40.79693  -73.93296        6.1   

          DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
453 2021-01-18  23.0             ,,N  2.70             ,,N  
454 2021-01-31  13.0             ,,N  0.15             ,,N  
455 2021-02-06   6.0             ,,N  1.59             ,,N  
456 2021-03-11  33.0            ,L,N  2.92            ,L,N  
465 2021-04-14  26.0             ,,N  3.87             ,,N  
466 2021-05-12  28.0             ,,N  3.34             ,,N  
483 2021-05-31   3.0             ,,N  2.96             ,,N  
487 2021-06-06   3.0             ,,N  0.81             ,,N  
491 2021-06-23  14.0             ,,N  0.45             ,,N  
493 2021-07-06  12.0             ,,N  2.30             ,,N  
494 2021-07-12   6.0             ,,N  5.67             ,,N  
495 2021-08-21  40.0            ,L,N  3.44            ,L,N  
496 2021-08-23   3.0             ,,N  4.45             ,,N  
507 2021-10-25  53.0            ,L,N  2.82            ,L,N  
508 2021-11-22  28.0             ,,N  5.25             ,,N

In [80]:
df[df["NAME"]=='NY CITY CENTRAL PARK, NY US']

STATION                         NAME  LATITUDE  LONGITUDE  \
9529  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9530  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9531  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9532  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9533  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
...           ...                          ...       ...        ...   
9889  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9890  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9891  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9892  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   
9893  USW00094728  NY CITY CENTRAL PARK, NY US  40.77898  -73.96925   

      ELEVATION       DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
9529       42.7 2021-01-01   NaN             NaN   NaN             NaN  
9530       42.7 2021-01-02   NaN             NaN   NaN             NaN  
9531       42.7 2021-01-03   NaN             NaN   NaN             NaN  
9532       42.7 2021-01-04   NaN             NaN   NaN             NaN  
9533       42.7 2021-01-05   NaN             NaN   NaN             NaN  
...         ...        ...   ...             ...   ...             ...  
9889       42.7 2021-12-27   NaN             NaN   NaN             NaN  
9890       42.7 2021-12-28   NaN             NaN   NaN             NaN  
9891       42.7 2021-12-29   NaN             NaN   NaN             NaN  
9892       42.7 2021-12-30   NaN             NaN   NaN             NaN  
9893       42.7 2021-12-31   NaN             NaN   NaN             NaN  

[365 rows x 10 columns]

In [12]:
weather_data1 = weather_data.groupby("DATE")["DAPR","MDPR"].count()

/var/folders/n8/dyf8zcj94f31mhgh7sfmdt5c0000gn/T/ipykernel_41458/1102332418.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  weather_data1 = weather_data.groupby("DATE")["DAPR","MDPR"].count()


In [13]:
weather_data1

DAPR  MDPR
DATE                  
2021-01-01     1     1
2021-01-02     0     0
2021-01-03     0     0
2021-01-04     0     0
2021-01-05     0     0
...          ...   ...
2021-12-27     0     0
2021-12-28     0     0
2021-12-29     0     0
2021-12-30     0     0
2021-12-31     0     0

[365 rows x 2 columns]

the Rainning Day is the 115/365.

In [14]:
rainday = (weather_data1['DAPR'] > 0 )
Rain_data = weather_data1.loc[rainday]
Rain_data.shape

(115, 2)

In [29]:
weather_data3 = weather_data.head(5)
weather_data3

STATION                   NAME  LATITUDE  LONGITUDE  ELEVATION  \
0  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
1  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
2  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
3  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
4  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   

        DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
0 2021-06-24   NaN             NaN   NaN             NaN  
1 2021-06-25   NaN             NaN   NaN             NaN  
2 2021-06-26   NaN             NaN   NaN             NaN  
3 2021-06-27   NaN             NaN   NaN             NaN  
4 2021-06-28   NaN             NaN   NaN             NaN

In [37]:
import geopy
import pandas as pd

geolocator = geopy.Nominatim(user_agent="check_1")

def get_zip_code(df,LATITUDE,LONGITUDE):
    #location = geolocator.reverse("{}, {}".format(x['LATITUDE_X'],x['LONGITUDE_X']))
    location = geolocator.reverse((df['LATITUDE'],df['LONGITUDE']))
    return location.raw['address']['postcode']

#weather_data3['zipcode'] = weather_data3.apply(get_zip_code, axis = 1)

weather_data3['zipcode'] = weather_data3.apply(get_zip_code, axis = 1,LATITUDE = weather_data3['LATITUDE'], LONGITUDE = weather_data3['LONGITUDE'])
weather_data3

/var/folders/n8/dyf8zcj94f31mhgh7sfmdt5c0000gn/T/ipykernel_41458/1162078318.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  weather_data3['zipcode'] = weather_data3.apply(get_zip_code, axis = 1,LATITUDE = weather_data3['LATITUDE'], LONGITUDE = weather_data3['LONGITUDE'])


STATION                   NAME  LATITUDE  LONGITUDE  ELEVATION  \
0  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
1  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
2  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
3  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
4  US1NJBG0043  MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   

        DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES zipcode  
0 2021-06-24   NaN             NaN   NaN             NaN   07607  
1 2021-06-25   NaN             NaN   NaN             NaN   07607  
2 2021-06-26   NaN             NaN   NaN             NaN   07607  
3 2021-06-27   NaN             NaN   NaN             NaN   07607  
4 2021-06-28   NaN             NaN   NaN             NaN   07607

In [38]:
NYC_weather_data = weather_data
NYC_weather_data

STATION                      NAME  LATITUDE  LONGITUDE  ELEVATION  \
0      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
1      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
2      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
3      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
4      US1NJBG0043     MAYWOOD 0.2 SW, NJ US  40.90001  -74.06641       24.7   
...            ...                       ...       ...        ...        ...   
11208  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11209  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11210  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11211  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   
11212  USW00094741  TETERBORO AIRPORT, NJ US  40.85898  -74.05616        0.8   

            DATE  DAPR DAPR_ATTRIBUTES  MDPR MDPR_ATTRIBUTES  
0     2021-06-24   NaN             NaN   NaN             NaN  
1     2021-06-25   NaN             NaN   NaN             NaN  
2     2021-06-26   NaN             NaN   NaN             NaN  
3     2021-06-27   NaN             NaN   NaN             NaN  
4     2021-06-28   NaN             NaN   NaN             NaN  
...          ...   ...             ...   ...             ...  
11208 2021-12-27   NaN             NaN   NaN             NaN  
11209 2021-12-28   NaN             NaN   NaN             NaN  
11210 2021-12-29   NaN             NaN   NaN             NaN  
11211 2021-12-30   NaN             NaN   NaN             NaN  
11212 2021-12-31   NaN             NaN   NaN             NaN  

[6992 rows x 10 columns]